Add calendar widgets (wibox and popup)

getzze opened the pull request:
There are two widgets: 
* wibox.widget displaying a month or a year.
The output is very similar to the UNIX `cal` command.
The code is based on the calendar implementations from blingbling and lain and it uses `wibox.widget.grid`. This widget doesn't have many properties, but it provides a wrapper property (`fn_embed`) to tweak all the individual cells.

* awful.widget is a wibox to display the calendar (year or month) as a popup. It has a lot of properties (too many maybe) but it is usable directly in `rc.lua`. Maybe this is too complicated to maintain.

Elv13 commented:
Oh nice, you will make a lot of people happy with this. I don't have time to do a full review right now, so I will just note some high level stuff.

 * There seems to be an issues found by `luacheck`
 * The "default" screenshot seem to have some extra unused pixels at the bottom
 * There is some line length issues in the code and example. For the grid layout I fixed those myself, but as I said, I currently don't have time for Awesome. Try to make sure the examples never create an horizontal scrollbar and keep the line length close to 80 char.
 * The coverage is already quite good (congratulation for that, this is appreciated), but could be a bit better. An integration test for the `awful.widget.calendar` should bump it in the green. Those are the files in `tests/test-*.lua`. You can run them locally using `run.sh`.
 * Your style system create cool looking widgets, but the API look a bit awkward. Could that callback be converted into a property for each "flag". Some kind of template properties?

I only reviewed some details of the documentation pages generated by your patch. I didn't check the code yet.

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,360 @@
+-- @treturn wibox The wibox calendar
+function calendar_popup:attach(widget, position)
+    local position = position or "tr"
+    widget:buttons(util.table.join(abutton({ }, 1, function ()
```
Change `util.table.join` to `gears.table.join`

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,360 @@
+                       opacity = args.opacity or 1,
+                       bg      = args.bg
+    }
+    util.table.crush(ret, calendar_popup, false)
```
Change `util.table.crush` to `gears.table.crush`

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,360 @@
+    }
+    ret:set_widget(widget)
+
+    ret:buttons(util.table.join(
```
Change `util.table.join` to `gears.table.join`

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,360 @@
+--
+--@DOC_wibox_awidget_calendar_year_wibox_EXAMPLE@
+--
+--    globalkeys = awful.util.table.join(globalkeys, awful.key({ modkey, "Control" }, "c",  function () year_calendar:toggle() end))
```
Change `awful.util.table.join` to `gears.table.join`

Also luacheck line too long (132 > 120)

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,360 @@
+local setmetatable = setmetatable
+local string = string
+local ipairs = ipairs
+local util = require("awful.util")
```
Remove ipairs and this line. Almost all of `awful.util` is deprecated.

Veratil commented on tests/examples/wibox/widget/calendar/week_numbers.lua:
```diff
@@ -0,0 +1,22 @@
+local wibox     = require("wibox") --DOC_HIDE
+local beautiful = require( "beautiful" ) --DOC_HIDE
+local lgi = require("lgi") --DOC_HIDE
+local Pango = lgi.Pango --DOC_HIDE
```
You can probably consolidate this and the previous line to:
`local Pango = require("lgi").Pango`
since lgi isn't used anywhere except here.

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,360 @@
+-- `opacity` and `bg` apply to the wibox itself, they are mainly useful to manage opacity
+-- by setting `opacity` for the false opacity or setting `bg="#00000000"` for compositor opacity.
+-- The `screen` argument forces the display of the wibox to this screen (instead of the focused screen by default).
+-- The `position` argument is a two-characters string describing the screen alignment "[vertical][horizontal]", e.g. "cc", "tr", "bl", ...
```
luacheck line too long (138 > 120)

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,360 @@
+
+local calendar_popup = { offset = 0, mt = {} }
+
+local properties = { "fg_color", "bg_color", "shape", "padding", "border_width", "border_color", "opacity" }
```
luacheck says this is unused

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,360 @@
+-- @treturn wibox The wibox calendar
+function calendar_popup:call_calendar(inc_offset, props)
+    local props = props or {}
+    local offset, position, s = inc_offset or 0, props.position or self.position, props.screen or self.screen or ascreen.focused()
```
luacheck line too long (130 > 120)

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,360 @@
+    end
+
+    local widget = self:get_widget()
+    local date, hint_width, hint_height
```
hint_width and hint_height are unused

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,360 @@
+-- @tparam[opt="tr"] string position Two characters string defining the position on the screen
+-- @treturn wibox The wibox calendar
+function calendar_popup:attach(widget, position)
+    local position = position or "tr"
```
luacheck is complaining about this, forgot what the fix was.

    lib/awful/widget/calendar.lua:271:11: variable position was previously defined as an argument on line 270

Elv13 replied:
This is shadowing the original `position`. Remove the `local` to fix it.

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,360 @@
+-- @treturn wibox A wibox containing the calendar
+local function get_cal_wibox(caltype, args)
+    if caltype ~= "month" and caltype ~= "year" then return end
+    local args = args or {}
```
Another luacheck complaint:

    lib/awful/widget/calendar.lua:290:11: variable args was previously defined as an argument on line 288

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,405 @@
+local function parse_date(date)
+    -- Split a string by a given `separator`
+    local function split(str, sep)
+        local sep, fields = sep or ":", {}
```
Another luacheck complaint:

    lib/wibox/widget/calendar.lua:113:15: variable sep was previously defined as an argument on line 112

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,405 @@
+    local column_fday  = (d.wday - d.day + 1 - week_start ) % 7
+
+    --local flags = {"header", "weekdays", "weeknumber", "normal", "focus"}
+    local t, i, j, w, flag, text, markup
```
luacheck: `markup` is unused

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,405 @@
+
+    -- Week days
+    i = 2
+    for j = start_column, num_columns do  -- columns
```
luacheck complaint:

    lib/wibox/widget/calendar.lua:184:9: shadowing definition of variable j on line 173

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,405 @@
+    i = start_row
+    j = column_fday + start_column
+    local current_week = nil
+    for d=1, month_days do
```
luacheck complaint:

    lib/wibox/widget/calendar.lua:196:9: shadowing definition of variable d on line 168

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,405 @@
+        forced_num_rows = 3,
+    }
+
+    local month_date = ""
```
luacheck complaint:

    lib/wibox/widget/calendar.lua:241:11: value assigned to variable month_date is unused

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,405 @@
+    return self._private.container or {}
+end
+
+function calendar:set_children(...)
```
luacheck complaint:

    lib/wibox/widget/calendar.lua:323:32: unused variable length argument

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,405 @@
+end
+
+-- Layout widget
+function calendar:layout(context, width, height)
```
luacheck complaint:

    lib/wibox/widget/calendar.lua:327:26: unused argument context

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,405 @@
+-- @tparam[opt="Monospace 10"] string font Font of the calendar
+-- @treturn widget The calendar widget
+local function get_calendar(t, date, font)
+    args = args or {}
```
luacheck complaint:

    lib/wibox/widget/calendar.lua:352:5: setting non-standard global variable args
    lib/wibox/widget/calendar.lua:352:12: accessing undefined variable args

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,405 @@
+    ret._private.week_numbers = beautiful.calendar_week_numbers or false
+    ret._private.start_sunday = beautiful.calendar_start_sunday or false
+    ret._private.hide_year    = t == "year" and true or false
+    ret._private.fn_embed     = function (w, flag) return w end
```
luacheck complaint:

    lib/wibox/widget/calendar.lua:368:46: unused argument flag

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,405 @@
+    args = args or {}
+    if t ~= "month" and t ~= "year" then return end
+    local ret = base.make_widget(nil, "calendar_"..t, {enable_properties = true})
+    util.table.crush(ret, calendar, true)
```
Change `util.table` to `gtable`

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,405 @@
+local setmetatable = setmetatable
+local string = string
+local ipairs = ipairs
+local util = require("awful.util")
```
Change to:
`local gtable = require("gears.table")`

Veratil commented on tests/examples/wibox/widget/calendar/month.lua:
```diff
@@ -0,0 +1,50 @@
+local wibox     = require("wibox") --DOC_HIDE
+local beautiful = require( "beautiful" ) --DOC_HIDE
+local lgi = require("lgi") --DOC_HIDE
+local Pango = lgi.Pango --DOC_HIDE
```
You can probably consolidate this and the previous line to:
`local Pango = require("lgi").Pango`
since lgi isn't used anywhere except here.

Veratil commented on tests/examples/wibox/widget/calendar/year.lua:
```diff
@@ -0,0 +1,24 @@
+local wibox     = require("wibox") --DOC_HIDE
+local beautiful = require( "beautiful" ) --DOC_HIDE
+local lgi = require("lgi") --DOC_HIDE
+local Pango = lgi.Pango --DOC_HIDE
```
You can probably consolidate this and the previous line to:
`local Pango = require("lgi").Pango`
since lgi isn't used anywhere except here.

Veratil commented on tests/examples/wibox/widget/defaults/calendar.lua:
```diff
@@ -0,0 +1,17 @@
+local wibox     = require("wibox") --DOC_HIDE
+local beautiful = require( "beautiful" ) --DOC_HIDE
+local lgi = require("lgi") --DOC_HIDE
+local Pango = lgi.Pango --DOC_HIDE
```
You can probably consolidate this and the previous line to:
`local Pango = require("lgi").Pango`
since lgi isn't used anywhere except here.

Veratil commented on tests/examples/wibox/widget/calendar/font.lua:
```diff
@@ -0,0 +1,18 @@
+local wibox     = require("wibox") --DOC_HIDE
+local beautiful = require( "beautiful" ) --DOC_HIDE
+local lgi = require("lgi") --DOC_HIDE
+local Pango = lgi.Pango --DOC_HIDE
```
You can probably consolidate this and the previous line to:
`local Pango = require("lgi").Pango`
since lgi isn't used anywhere except here.

Veratil commented on tests/examples/wibox/widget/calendar/fn_embed_cell.lua:
```diff
@@ -0,0 +1,67 @@
+local gears     = require("gears") --DOC_HIDE
+local beautiful = require( "beautiful" ) --DOC_HIDE
+local lgi = require("lgi") --DOC_HIDE
+local Pango = lgi.Pango --DOC_HIDE
```
You can probably consolidate this and the previous line to:
`local Pango = require("lgi").Pango`
since lgi isn't used anywhere except here.

getzze commented:
I corrected the `luacheck` errors and added a `test-awful-widget-calendar.lua` for the awful.widget.
 @Elv13:
* the calendar has a fixed size of 8 rows, so for some months the bottom row (can even be two rows) is empty.
* The properties are a problem. First, I had put all the properties of the `awful.widget` in the `wibox.widget`, but that is a lot (8 flag x 8 properties ...). So I came up with the `fn_embed` callback which is quite handy because it gives a lot of liberty. However it is slower, displaying the year calendar makes Awesome lag.
If somebody has a suggestion to deal with the properties...

Elv13 commented:
Thanks for the stellar code coverage! For one of the current CI error, you should squash some of your commits back to "add a cal"  and "test the cal". For the other, I didn't investigate, but it looks like it found something:

```
== Running /home/travis/build/awesomeWM/awesome/tests/test-awful-widget-calendar.lua ==

2017-03-31 23:19:25 W: Cannot reliably detect EOF on an GIOInputStream with this LGI version

Error: running function for step 1/5 (@2): [string "--- Test for awful.widget.calendar..."]:43: assertion failed!

stack traceback:

	[C]: in function 'assert'

	[string "--- Test for awful.widget.calendar..."]:43: in function <[string "--- Test for awful.widget.calendar..."]:36>

	(...tail calls...)

	[C]: in function 'xpcall'

	/home/travis/build/awesomeWM/awesome/tests/_runner.lua:48: in function </home/travis/build/awesomeWM/awesome/tests/_runner.lua:42>

	[C]: in function 'xpcall'

	lib/gears/protected_call.lua:36: in function <lib/gears/protected_call.lua:35>

	(...tail calls...)

	lib/gears/timer.lua:226: in function <lib/gears/timer.lua:224>!

===> ERROR running /home/travis/build/awesomeWM/awesome/tests/test-awful-widget-calendar.lua <===

Error: running function for step 1/5 (@2): [string "--- Test for awful.widget.calendar..."]:43: assertion failed!
```

Veratil commented on lib/awful/widget/calendar.lua:
```diff
@@ -0,0 +1,381 @@
+
+local setmetatable = setmetatable
+local string = string
+local gtable = require("gears.table")
```
You already import gears below, you could remove this line and expand `gtable` to `gears.table`

Veratil replied:
Actually looking through everything, I think you only use `gears.table` and `gears.shape`.

Veratil commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,401 @@
+-- @treturn {number,number,number} List of day, month, year (nil if undefined)
+local function parse_date(date)
+    -- Split a string by a given `separator`
+    local function split(str, sep)
```
[gears.string.split](https://awesomewm.org/apidoc/libraries/gears.string.html#split)

Veratil replied:
As a note, if gsub is faster than gmatch then you could update the [split function here](https://github.com/awesomeWM/awesome/blob/master/lib/gears/string.lua#L95). Just make sure it still functions the same where it's used in `gears.filesystem` and anywhere else.

Elv13 commented:
@actionless @psychon @blueyed Can someone else review this, I really have no time

@getzze Really sorry about the delay here, this is very good work

psychon commented:
@Elv13 Sorry, but I also seem to fail to find the time to do a review. All I figured out so far is that this does not seem to spawn external processes to do the work (good!) and that there is a split into `wibox.widget.calender` and `awful.widget.calender`, but I haven't even figured out the difference between the two (might make sense to rename one of them so that the name "calender" is unique).

psychon commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,401 @@
+-- Parse a date for the day, month and year
+-- @tparam string date Date in the "%d %m %Y" format
+-- @treturn {number,number,number} List of day, month, year (nil if undefined)
+local function parse_date(date)
```
How about changing the API so that users already provide `day, month, year` so that no such string parsing is needed?

getzze replied:
`date` is a single property, otherwise there are three. but maybe a table would be ok.

psychon replied:
I don't have much of a problem with three properties either, but sure, I don't mind a table either. Just string parsing should be avoided where possible. ;-)

psychon commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,401 @@
+    i = 2
+    for jj = start_column, num_columns do  -- columns
+        flag = "weekdays"
+        t = os.time{year=2006, month=1, day=jj-start_column+week_start}
```
What's so special about january 2006? Why does this not use the actual year and month that were provided?

getzze replied:
I used the code from another calendar widget, but the creation of the 'weekdays' cells can be done in the next for-loop indeed.

psychon replied:
> I used the code from another calendar widget,

Copyright-wise, could this be a problem? Even if that other implementation is GPL'd (is it?), then this would violate the attribution clause in the GPL.

getzze replied:
I took inspiration - not exact copy - from this one:
https://github.com/deficient/calendar
Otherwise, for calendar_popup (awful widget), the code is very similar to lain.calendar.
These two should be mentioned I guess.

psychon commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,401 @@
+    for jj = start_column, num_columns do  -- columns
+        flag = "weekdays"
+        t = os.time{year=2006, month=1, day=jj-start_column+week_start}
+        text = string.sub(os.date("%a", t), 1, 2)
```
Why do you abbreviate the abbreviated weekday name? AFAIR there are locales where this makes the day non-unique.

getzze replied:
This is the behavior of the `cal` command-line in linux, so the cell size is two characters, like for the days numbers. I didn't find another way to ouput a two-characters weekday with the LUA libraries.

psychon replied:
Hm... I am not sure `cal` should do that either...

actionless replied:
i would use max length of abbreviated weekday name as a column width (for all columns)

getzze replied:
I will add an option to show three characters

psychon commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,401 @@
+    gtable.crush(ret, calendar, true)
+
+    ret.type = t
+    ret._private.calendar_cache = {}
```
Why does this have a cache at all? I am not convinced that something that displays a calender has to cache things.
(Also, I don't think that the cache should be keyed via a string, but let's ignore this)

getzze replied:
Creating a year calendar is slow, so the cache helps if you want to check the next year and come back to the current year. Also I didn't understand how the widget cache works so I did this.

psychon replied:
The widget cache in `wibox.widget.base`? It just caches the result of `:fit()` and `:layout()` and is cleared when the respective signals are emitted. So that would not help here.

psychon commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,401 @@
+
+    -- default values
+    ret._private.date = date
+    ret._private.font = font or beautiful.calendar_font or "Monospace 10"
```
Why the hardcoded `Monospace 10`? Just use `nil` as the default.

actionless replied:
i would add just `beautiful.font` before Monospace but i think monospace is good one, because if not unset it could use sans and make calendar unreadable

getzze replied:
Because of the grid alignment, a monospace font is nicer. But it works also with other fonts, so as you want

psychon commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,401 @@
+    return get_calendar(caltype, ...)
+end
+
+return setmetatable(calendar, calendar.mt)
```
Our usual vim modeline is missing.

psychon commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,401 @@
+    if self._private.container then
+        return { base.place_widget_at(self._private.container, 0, 0, width, height) }
+    end
+    return {}
```
This widget just wraps another widget. As I see it, that other widget is a `grid` layout. How about just creating the grid up-front and passing it as the first argument to `base.make_widget`, so that the `:fit` and `:layout` methods become unnecessary here.

getzze replied:
There were two problems with this solution:
1-the api of the `grid` widget was exposed as the api of the calendar widget, so names could collide.
2-to change the date of the calendar widget, you have to destroy the last grid and create a new one (I think) and it is not so straightforward to have a cache (see other comment).

psychon replied:
1) I don't mean to replace `local my_widget = base.create_widget() my_widget._private.real_widget  = grid()` with `local my_widget = grid()`. Instead I mean: Replace this with: `local real_widget = grid() local my_widget = base.create_widget(real_widget) my_widget._private.real_widget = real_widget`. That way you do not have to implement `:fit()` and `:layout` at all, but instead they are automatically redirected to the target widget. The only problem is that you then cannot replace the grid layout any more and always have to use the same. However, that should not be a problem because of (2):
(2): There is a `:reset()` method on basically all layouts, also on the `grid` layout. That way you do not have to replace the grid.

And for "not so straightforward to have a cache": Well, I am not that much of a fan of this cache so far anyway. ;-)

getzze replied:
Actually, `real_widget` is not necessarily a `grid`. It can be embedded in another widget, see the last lines of the `create_month` and `create_year` functions.
The code for the `:fit` and `:layout` functions are the same as for others `wibox.widget.container`, it does not require much maintenance.

psychon commented on lib/wibox/widget/calendar.lua:
```diff
@@ -0,0 +1,401 @@
+-- @tparam[opt="Monospace 10"] string font Font of the calendar
+-- @treturn widget The calendar widget
+local function get_calendar(t, date, font)
+    if t ~= "month" and t ~= "year" then return end
```
Replace `return` by an `error()` call with a useful error message. That makes debugging a lot easier, if this ever goes wrong.

getzze replied:
maybe I could remove the test and make the function completely local (by removing calendar.mt:__call(caltype, ...)).

getzze commented:
Different changes:
 - rename awful.widget.calendar -> calendar_popup
 - use table {year=..., month=..., day=...} instead of strings for the date
 - rename the flag `weekdays` to `weekday`
 - add `long_weekdays` property to display the weekdays with 3 chars instead of 2
 - remove the cache
 - use a container.background as proxy for the calendar widget, to be able to remove the `:fit` and `:layout` functions.
 - remove the global calendar(type, ...) function
 - add a `date` argument to `fn_embed` function so you can decorate the days depending on the date (change colors for weekend days, add a list of bank holidays...)

About the default font, should I leave "Monospace"?
I also realized about a non-consistent behavior between  calendar.year and calendar.month: the year calendar always highlights the current day, whereas calendar.month highlights the specified day. It is the same behavior as the cal command-line. I thought about changing it by adding a `no_highlight` property. If it is set, no day is highlighted, otherwise, if a day is specified, it is the one that is highlighted, otherwise the current day is highlighted. But it's maybe a bit complicated.

Elv13 commented:
> About the default font, should I leave "Monospace"?

I think so, it would otherwise look ugly.

Also, could you rebase on top of master so the CI shows up again? (it did run, but isn't visible)

psychon commented:
@Elv13 How unlikely is it that you can find the time to look at this in the near future?

Elv13 commented:
@psychon I built it, tested it and looked at the code and API. I think for a first version, this is getting good enough to merge. One way or another there is no way this could have been built in 10 lines of code, so the size and sometime non-trivial code are to be expected. I am fine with merging this as-is (once luacheck is fixed)

@getzze Thanks for this. Do you want this to be merged? I can fix the luacheck regression when I merge this.

Elv13 commented:
Thanks a lot to everybody who contributed to this PR, be it code, inspiration, testing or reviews.
